src/sage/graphs/graph_decompositions/slice_decomposition.pyx

```diff
@@ -0,0 +1,970 @@
+            # create a new part using subpart
+            if subpart[p] < old_nparts:
+                subpart[p] = nparts
+                part_head[nparts] = j
+                if xslice_len != NULL:
+                    part_len[nparts] = 0
+                subpart[nparts] = 0
+                nparts += 1
+
+            # Finally, we update the name of the part for position j and set v
+            # as predecessor of u
+            part_of[j] = subpart[p]
+            if xslice_len != NULL:
+                part_len[p] -= 1
+                part_len[subpart[p]] += 1
+            if pred != NULL:
+                deref(pred)[j] = i
+
+    if need_to_delete_sigma_inv:
+        del sigma_inv
+
+def slice_decomposition(G, initial_vertex=None):
+    r"""
+    Compute a slice decomposition of the simple undirected graph
+
+    INPUT:
+
+    - ``G`` -- a Sage graph.
+
+    - ``initial_vertex`` -- (default: ``None``); the first vertex to consider.
+
+    OUTPUT:
+
+    An object of type :class:`~sage.graphs.graph_decompositions.slice_decomposition.SliceDecomposition`
+    that represents a slice decomposition of ``G``
+
+    .. NOTE::
+
+        Loops and multiple edges are ignored during the computation of the slice
+        decomposition.
+
+    ALGORITHM:
+
+    The method use the algorithm based on "partition refinement" described in
+    [HMPV2000]_ and [TCHP2008]_.
+    The time complexity of this algorithm is in `O(n + m)`, and our
+    implementation follows that complexity for ``SparseGraph``. For
+    ``DenseGraph``, the complexity is `O(n^2)`.
+
+    EXAMPLES:
+
+    Slice decomposition of the Petersen Graph::
+
+        sage: G = graphs.PetersenGraph()
+        sage: SD = G.slice_decomposition(); SD
+        [0[1[4[5]]][2[6]][3][9][7][8]]
+
+    The graph can have loops or multiple edges but they are ignored::
+
+        sage: H = Graph(G,loops=True,multiedges=True)
+        sage: H.add_edges([(4, 4), (2, 2), (1, 6)])
+        sage: SD2 = H.slice_decomposition()
+        sage: SD2 == SD
+        True
+        sage: SD2.underlying_graph() == G.to_simple(immutable=True)
+        True
+
+    The tree corresponding to the slice decomposition can be displayed using
+    ``view``::
+
+        sage: from sage.graphs.graph_latex import check_tkz_graph
+        sage: check_tkz_graph() # random - depends on Tex installation
+        sage: view(G) # not tested
+        sage: latex(G) # to obtain the corresponding LaTeX code
+        \begin{tikzpicture}
+        ...
+        \end{tikzpicture}
+
+    Slice decompositions are only defined for undirected graphs::
+
+        sage: from sage.graphs.graph_decompositions.slice_decomposition import slice_decomposition
+        sage: slice_decomposition(DiGraph())
+        Traceback (most recent call last):
+        ...
+        ValueError: Graph must be undirected
+
+    """
+    return SliceDecomposition(G, initial_vertex=initial_vertex)
+
+
+cdef class SliceDecomposition(SageObject):
+
```

Comment:
Add a description of the class.

Reply:
I though that a class docstring was not mandatory, because if it does not exists, the docstring from `__init__` is used.
Should I just write a one line description or more thorough description ?

Reply:
See for instance https://doc.sagemath.org/html/en/reference/graphs/sage/graphs/views.html
The documentation of the class (and the doctests) are in the online documentation. Internal methods such as `__init__` are not.

I'm not sure of what is mandatory, but it's nice to have documentation in the reference manual.

Reply:
The docstring from `__init__` does appear in the [preview of the reference manual](https://preview-38299--sagemath.netlify.app/html/en/reference/graphs/sage/graphs/graph_decompositions/slice_decomposition#sage.graphs.graph_decompositions.slice_decomposition.SliceDecomposition) as the docstring of the class itself.
It also appear in interactive Sage using `sage.graphs.graph_decompositions.slice_decomposition.SliceDecomposition?`

Reply:
I didn't know that. So then, you can simply add relevant information and doctests to the `__init__` method.

Reply:
I added some more doc in commit [3b428a6](https://github.com/sagemath/sage/pull/38299/commits/3b428a6cce9f72b22677be8a2262f4ac4963ec92)